src/utils/generator.js

```diff
@@ -197,6 +197,21 @@ export class Generator {
     );
   }
 
+  emitDoWhileStatement(test: AbstractValue, body: Generator) {
+    this._addEntry({
+      args: [],
+      buildNode: function([], context: SerializationContext) {
+        let statements = context.serializeGenerator(body);
+        let testId = test.intrinsicName;
+        invariant(testId !== undefined);
```

Comment:
Not all abstract values have an intrinsic name. Why is this guaranteed?

Reply:
Related, I suspect you were trying hard to avoid using the regular mechanism to turn values into nodes, as this `test` thingy is something very special and not a regular value. Scary implicit assumptions... I feel like we need to do some more fundamental distinction to avoid chaos down the road.

Reply:
The intrinsic name requirement is a precondition of emitDoWhileStatement. I would hoist it, but then Flow complains. Repeating it seemed a bit over the top to me. I'll add a comment.